pkg/filestore/s3/s3.go

```diff
@@ -0,0 +1,182 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/ec2rolecreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/filestore"
+)
+
+type Store struct {
+	client *s3.S3
+	bucket string
+
+	logger *zap.Logger
+}
+
+type Option func(*Store)
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(s *Store) {
+		s.logger = logger.Named("s3")
+	}
+}
+
+func NewStore(region, profile, credentialsFile, bucket string, opts ...Option) (*Store, error) {
+	if region == "" {
+		return nil, fmt.Errorf("region is required field")
+	}
+	if bucket == "" {
+		return nil, fmt.Errorf("bucket is required field")
+	}
+
+	s := &Store{
+		bucket: bucket,
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a session: %w", err)
+	}
+	creds := credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&credentials.EnvProvider{},
+			&credentials.SharedCredentialsProvider{
+				Filename: credentialsFile,
+				Profile:  profile,
+			},
+			&ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(sess),
+			},
+		},
+	)
+	cfg := aws.NewConfig().WithRegion(region).WithCredentials(creds)
+	s.client = s3.New(sess, cfg)
+
+	return s, nil
+}
+
+func (s *Store) NewReader(ctx context.Context, path string) (rc io.ReadCloser, err error) {
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(path),
+	}
+	out, err := s.client.GetObjectWithContext(ctx, input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchKey:
+				err = filestore.ErrNotFound
+				return
+			case s3.ErrCodeInvalidObjectState:
+				err = fmt.Errorf("invalid object state: %w", err)
+				return
+			default:
+				err = fmt.Errorf("unexpected aws error given: %w", err)
+				return
+			}
+		}
+		err = fmt.Errorf("unknown error given: %w", err)
+		return
+	}
+	return out.Body, nil
+}
+
+func (s *Store) GetObject(ctx context.Context, path string) (object filestore.Object, err error) {
+	object.Path = path
+	rc, err := s.NewReader(ctx, path)
+	if err != nil {
+		return
+	}
+	content, err := ioutil.ReadAll(rc)
+	if err != nil {
+		rc.Close()
```

Comment:
I feel just enclosing defer statement is better. I'm aware that GCS's one is the same. But never mind, I'd say that should be so as well.

```go
	defer func() {
		if err := rc.Close(); err != nil {
			s.logger.Error("failed to close...", zap.Error(err))
		}
	}()
```

Reply:
Addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/95573edcdacb1d3a5512c7cb0c5952d9eb9ca0a6 🙏 PTAL 👀